### PR TITLE
feat: allow changing backoff and limiter per request

### PR DESCRIPTION
**Pull request by shoom3301**
Context: https://cowservices.slack.com/archives/C0361CDG8GP/p1715853690864799?thread_ts=1715851334.267859&cid=C0361CDG8GP

In some cases we need to configure backoff or limiter per request.
For example, by default we want to use default strategy (10 retries), but for a specific case we want to use 1 retry.

To allow that I've included `RequestOptions` into `ApiContext` and added override of the parameters.

### CODE REVIEW

File: src/order-book/api.ts [hunk 1]
```diff
@@ -126,15 +125,15 @@ export type GetOrdersRequest = {
  * @see {@link OrderBook API https://github.com/cowprotocol/services}
  */
 export class OrderBookApi {
-  public context: ApiContext & RequestOptions
```
Comment (alfetopito) [inline src/order-book/api.ts:129]: Why remove this?

Comment (shoom3301) [reply src/order-book/api.ts:131]: `RequestOptions` was merged into `ApiContext`

Comment (shoom3301) [reply src/order-book/api.ts:131]: image

File: src/common/configs.ts [hunk 2]
```diff
@@ -67,6 +67,8 @@ export interface ApiContext {
   chainId: SupportedChainId
   env: CowEnv
   baseUrls?: ApiBaseUrls
+  limiterOpts?: RateLimiterOpts
```
Comment (alfetopito) [inline src/common/configs.ts:70]: This is added here but not used anywhere in the PR.
Is it need?

**Review by anxolin — APPROVED** [review]

File: src/common/configs.ts (hunk 2 above)
Comment (shoom3301) [reply src/common/configs.ts:70]: image

It is used here

Comment (alfetopito) [reply src/common/configs.ts:70]: But this wasn't introduced in this PR.
Was it just not properly typed before?

Comment (shoom3301) [reply src/common/configs.ts:70]: Sorry, good point actually!

Added a fix: https://github.com/cowprotocol/cow-sdk/pull/208/commits/826f4101b648c07b9d03f06d60772ed1bfa8ba1b

File: src/order-book/api.ts [hunk 3]
```diff
@@ -410,10 +409,11 @@ export class OrderBookApi {
    * @returns The response from the API.
    */
   private fetch<T>(params: FetchParams, contextOverride: PartialApiContext = {}): Promise<T> {
-    const { chainId, env } = this.getContextWithOverride(contextOverride)
+    const { chainId, env, backoffOpts: _backoffOpts } = this.getContextWithOverride(contextOverride)
     const baseUrl = this.getApiBaseUrls(env)[chainId]
-    const backoffOpts = this.context.backoffOpts || DEFAULT_BACKOFF_OPTIONS
+    const backoffOpts = _backoffOpts || DEFAULT_BACKOFF_OPTIONS
+    const rateLimiter = contextOverride.limiterOpts ? new RateLimiter(contextOverride.limiterOpts) : this.rateLimiter
```
Comment (alfetopito) [inline src/order-book/api.ts:415]: Doesn't this need to be fetched from `this.getContextWithOverride(contextOverride)` like `backoffOpts`?

Comment (shoom3301) [reply src/order-book/api.ts:415]: No, here we use `this.rateLimiter` by default and create a new only only when `limiterOpts` is set in `contextOverride`

**Pull request merged**